Document dbtRunner (programmatic invocation) (#3118)
fishtown-analytics/docs.getdbt.com: website/docs/reference/programmatic-invocations.md

---
title: "Programmatic invocations"
---

In v1.5, dbt-core added support for programmatic invocations. The intent is to expose the existing dbt CLI via a Python entry point, such that top-level commands are callable from within a Python script or application.

The entry point is a `dbtRunner` class, which allows you to `invoke` the same commands as on the CLI.

```python
from dbt.cli.main import dbtRunner, dbtRunnerResult

# initialize
dbt = dbtRunner()

# create CLI args as a list of strings
cli_args = ["run", "--select", "tag:my_tag"]

# run the command
res: dbtRunnerResult = dbt.invoke(cli_args)

# inspect the results
for r in res.result:
    print(f"{r.node.name}: {r.status}")

> ```

Reviewer: This example is verified to work for me.

## Advanced usage patterns

:::caution
The syntax and support for these patterns are liable to change in future versions of `dbt-core`.
:::

The goal of `dbtRunner` is to offer parity with CLI workflows, within a programmatic environment. There are a few advanced usage patterns that extend what's possible with the CLI.

### Reusing objects

Pass pre-constructed objects into `dbtRunner`, to avoid recreating those objects by reading files from disk. Currently, the only object supported is the `Manifest` (project contents).

```python
from dbt.cli.main import dbtRunner, dbtRunnerResult
from dbt.contracts.graph.manifest import Manifest

# use 'parse' command to load a Manifest
res: dbtRunnerResult = dbtRunner().invoke(["parse"])
manifest: Manifest = res.result

# introspect manifest
# e.g. assert every public model has a description
for node in manifest.nodes.values():
    if node.resource_type == "model" and node.access == "public":

>         assert node.description != ""

Reviewer: ```suggestion
        assert node.description != "", f"{node.name} is missing a description"
```

# reuse this manifest in subsequent commands to skip parsing
dbt = dbtRunner(manifest=manifest)
cli_args = ["run", "--select", "tag:my_tag"]
res = dbt.invoke(cli_args)

> ```

Reviewer: This example is verified to work for me.

Using the following assertion made it more clear which model is missing a `description`:
```
        assert node.description != "", f"{node.name} is missing a description"
```

Author: @dbeatty10 That is nicer! Do you want to suggest it, or commit directly? Mi PR es tu PR

### Registering callbacks

Register `callbacks` on dbt's `EventManager`, to access structured events and enable custom logging. The current behavior of callbacks is to block subsequent steps from proceeding; this functionality is not guaranteed in future versions.

```python
from dbt.cli.main import dbtRunner
from dbt.events.base_types import EventMsg

def print_version_callback(event: EventMsg):
    if event.info.name == "MainReportVersion":
        print(f"We are thrilled to be running dbt{event.data.version}")

dbt = dbtRunner(callbacks=[print_version_callback])
dbt.invoke(["list"])

> ```

Reviewer: We _are_ thrilled to be running dbt=1.5.0-b5! Confirming that this example worked for me 👍

### Overriding parameters

Pass in parameters as Pythonic `**kwargs`, instead of a list of CLI-style strings. At present, dbt will not do any validation or type coercion on your inputs. The subcommand must be specified, in a list, as the first positional argument.
```python
from dbt.cli.main import dbtRunner
dbt = dbtRunner()

# these are equivalent

> dbt.invoke(["--fail-fast", "run", "--select", "tag:my_tag")

Reviewer: There's a little `]` running as fast as it can trying to catch the v1.5 bus...

🚌 . . . . . . . . 🏃 

Will it make it?!

```suggestion
dbt.invoke(["--fail-fast", "run", "--select", "tag:my_tag"])
```

Author: :)) nice catch!

---
title: "Programmatic invocations"
---

> In v1.5, dbt-core added support for a first-ever Python API for programmatic invocations. The intent of this API is provide **exact parity** with CLI functionality, callable from within a Python script or application.

Reviewer 2: ```suggestion
In v1.5, dbt-core added support for a first-ever Python API for programmatic invocations. The intent of this API is to provide **exact parity** with CLI functionality, callable from within a Python script or application.
```

> In v1.5, dbt-core added support for a first-ever Python API for programmatic invocations. The intent of this API is to provide **exact parity** with CLI functionality, callable from within a Python script or application.

Reviewer 3: IDK if we should call this functionality an `API`.

Author: Yeah, I get that. What language would you prefer to see here? "Entry point"?

> The entrypoint is a `dbtRunner` class that wraps around `dbt-core`'s CLI, and allows you to "invoke" CLI commands as Python methods. Each command returns a tuple of:

Reviewer 2: ```suggestion
The entry point is a `dbtRunner` class that wraps around the `dbt-core` CLI and allows you to "invoke" CLI commands as Python methods. Each command returns a tuple of:
```

- `results` (return type depends on command)
- `success` (boolean): whether the command succeeded

```python
from dbt.cli.main import dbtRunner

# initialize
dbt = dbtRunner()

# create CLI args as a list of strings
cli_args = ['run', '--select', 'some_models']

# run the command

> results: List[Any]
> success: boolean

Reviewer: These need to be commented out for the script to run. Other than this, the example works out-of-the-box.
```suggestion
# results: List[Any]
# success: boolean
```

Author: I updated the example here now that `dbtRunner` returns `dbtRunnerResult`, instead of a tuple (changed in https://github.com/dbt-labs/dbt-core/pull/7242)

results, success = dbt.invoke(cli_args)
```

### Caveats

> We are not committed to this interface forever. We do expect it to unlock real use cases in the short & medium term, while we work on a more stable long-term interfaces into dbt-core as a mature library & Python SDK.

Reviewer 2: ```suggestion
We are not committed to this interface forever. We expect it to unlock real use cases in the short & medium term while we work on a more stable long-term interface into dbt-core as a mature library & Python SDK.
```

- The `results` objects returned by each command are liable to change. Some of these objects are already contracted, in part, because they overlap with [dbt artifacts](dbt-artifacts).

### Advanced usage patterns

Pass pre-constructed objects into `dbtRunner`, to avoid recreating those objects by reading files from disk.

```python
import os
from dbt.cli.main import dbtRunner

# load profile + project
project_dir = os.getcwd()
profile = load_profile(project_dir, {}, 'testing-postgres')
project = load_project(project_dir, False, profile, {})
  
# initialize the runner with pre-loaded profile and project
dbt = dbtRunner(profile=profile, project=project)

# use 'parse' command to load a Manifest
results, success = dbt.invoke(['parse'])
# TODO get manifest from parse results

dbt = dbtRunner(profile=profile, project=project, manifest=manifest)

# run more commands, re-using the components saved above
res, success = dbt.invoke(cli_args)
```

Register a `callback` on the `EventManager`, to access structured events and enable custom logging.
```
TODO
```

Pass in params as `**kwargs`, instead of parsing from a list of strings (CLI-style).
```

> TODO

Reviewer 3: This will ship as part of https://github.com/dbt-labs/dbt-core/pull/7274, can pull an example from tests.

Author: thank you!

---
title: "Programmatic invocations"
---

> In v1.5, dbt-core added support for programmatic invocations. The intent of this entry point is provide **exact parity** with CLI functionality, callable from within a Python script or application.

Reviewer 3: Something like this?

> The intent of this entry point is to expose the existing dbt CLI via a Python entrypoint, so it is callable from a Python script or application.

> The main entry point is a `dbtRunner` class that wraps around `dbt-core`'s CLI, and allows you to "invoke" CLI commands as Python methods. Each command returns a `dbtRunnerResult` object, which has three attributes:

Reviewer 3: This should be the opposite? The CLI technically wraps the Click runner that `dbtRunner` depends on.

- `success` (bool): Whether the command succeeded.
- `result`: If the command completed (successfully or with handled errors), its result(s). Return type varies by command.
- `exception`: If the dbt invocation encountered an unhandled error and did not complete, the exception it encountered.

```python
from dbt.cli.main import dbtRunner, dbtRunnerResult

# initialize
dbt = dbtRunner()

# create CLI args as a list of strings
cli_args = ["run", "--select", "tag:my_tag"]

# run the command
res: dbtRunnerResult = dbt.invoke(cli_args)

# inspect the results
for r in res.result:
    print(f"{r.node.name}: {r.status}")
```

### `dbtRunnerResult` and exit codes

> There is a 1:1 correspondence between [CLI exit codes](reference/exit-codes) and the `dbtRunnerResult` returned by a programmatic invocation:

Reviewer 3: Let's put this in a table?

Reviewer: @aranke I think the table will be covered by the link.

i.e., `[CLI exit codes](reference/exit-codes)` will link to [this](https://docs.getdbt.com/reference/exit-codes) page which has the following table:

<img width="550" alt="image" src="https://user-images.githubusercontent.com/44704949/231562379-22d93e29-ee0a-43b7-b353-50d29e3b7865.png">

Author: I like the idea of tabling it up here as well!

Reviewer 3: Let's copy the same table here; most people aren't going to click on links.

Reviewer 3: Wait, it's a different table – here the columns should be: `scenario`, `success`, `exit_code`, `result`, `exception`.

1. Completed without error: `success=True`, exit code `0`, `result` populated based on the command, `exception` empty.
2. At least one handled error (e.g. test failure, model build error): `success=False`, exit code `1`, `result` populated (based on the command), `exception` empty.
3. Unhandled error: `success=False`, exit code `2`, `exception` populated with top-level error, `result` empty because the invocation did not gracefully complete and return results.

> ### Caveats

Reviewer 3: Should this be SLA or similar?

Author: As in - our SLA (set of guarantees) to end users, in terms of stability of the interface across future versions?

Reviewer 3: Yes, correct.

    print(f"{r.node.name}: {r.status}")
```

## `dbtRunnerResult`

Each command returns a `dbtRunnerResult` object, which has three attributes:
- `success` (bool): Whether the command succeeded.
- `result`: If the command completed (successfully or with handled errors), its result(s). Return type varies by command.
- `exception`: If the dbt invocation encountered an unhandled error and did not complete, the exception it encountered.

There is a 1:1 correspondence between [CLI exit codes](reference/exit-codes) and the `dbtRunnerResult` returned by a programmatic invocation:

> | Scenario                                                                                    | CLI Exit Code | `success` | `result`         | `exception` |

Reviewer 3: Nit: numbers should be right-aligned using the `|---:|` header.

Author: TIL!

|---------------------------------------------------------------------------------------------|---------------|-----------|-------------------|-------------|
| Invocation completed without error                                                          | 0             | `True`      | varies by command | `None`        |
| Invocation completed with at least one handled error (e.g. test failure, model build error) | 1             | `False`     | varies by command | `None`        |
| Unhandled error. Invocation did not complete, and returns no results.                       | 2             | `False`     | `None`              | Exception   |

## Commitments & Caveats

> From dbt Core v1.5 onward, we making an ongoing commitment to providing a Python entry point at functional parity with dbt-core's CLI. We reserve the right to change the underlying implementation used to achieve that goal. We expect that the current implementation will unlock real use cases, in the short & medium term, while we work on a set of stable, long-term interfaces that will ultimately replace it.

Reviewer 3: I like this language, ty.

dbt.invoke(["list"])
```

### Overriding parameters

> Pass in parameters as Pythonic `**kwargs`, instead of a list of CLI-style strings. At present, dbt will not do any validation or type coercion on your inputs. The subcommand must be specified, in a list, as the first positional argument.

Reviewer 3: Nit: this is not Pythonic style, Python lets you do this as an escape hatch.

Author: Heh, good point. I'm just going to change this to `Pass in parameters as keyword arguments`
